src/main/java/com/exadel/discountwebapp/discount/service/DiscountService.java

```diff
@@ -48,15 +55,40 @@ public DiscountResponseVO create(DiscountRequestVO request) {
 
     @Transactional(propagation = Propagation.REQUIRED)
     public DiscountResponseVO update(Long id, DiscountRequestVO request) {
-        Discount discount = discountRepository.findById(id)
-                .orElseThrow(() -> new EntityNotFoundException(Discount.class, "id", id));
+        discountValidator.checkIfDiscountIsPresentById(id);
+        Discount discount = discountRepository.findById(id).get();
         discountMapper.updateEntity(request, discount);
         discountRepository.save(discount);
         return discountMapper.toVO(discount);
     }
 
     @Transactional(propagation = Propagation.REQUIRED)
     public void deleteById(Long id) {
+        discountValidator.checkIfDiscountIsPresentById(id);
         discountRepository.deleteById(id);
     }
+
+    @Transactional(propagation = Propagation.REQUIRED)
+    public ResponseEntity<String> addDiscountToFavorites(Long userId, Long discountId) {
+        discountValidator.checkIfDiscountIsPresentById(discountId);
+        Discount discount = discountRepository.findById(discountId).get();
+        User user = userRepository.findById(userId)
+                .orElseThrow(() -> new EntityNotFoundException(User.class, "id", userId));
+        discountValidator.checkFavoritesToAdd(discount, user);
+        discount.getFavoriteUsers().add(user);
+        discountRepository.save(discount);
+        return new ResponseEntity<>(HttpStatus.OK);
+    }
+
+    @Transactional(propagation = Propagation.REQUIRED)
+    public ResponseEntity<String> deleteDiscountFromFavorites(Long userId, Long discountId) {
+        discountValidator.checkIfDiscountIsPresentById(discountId);
```

Comment:
no need to check if discount was present to delete it. 
If there were no discount then we already have a result that we want.